pkg/metrics/middleware.go

```diff
@@ -53,15 +53,17 @@ func getShortPath(s string) string {
 // ServeHTTP implements the http.Handler interface. It records the timestamp
 // this API call began at, then chains to the next handler.
 func (m *APIEventTSHelper) ServeHTTP(r http.ResponseWriter, req *http.Request) {
-	if req != nil {
-		m.TSGauge.WithLabelValues(LabelEventSourceAPI, req.URL.Path, req.Method)
-	}
 	duration := spanstat.Start()
 	rw := &responderWrapper{ResponseWriter: r}
 	m.Next.ServeHTTP(rw, req)
 	if req != nil && req.URL != nil && req.URL.Path != "" {
 		path := getShortPath(req.URL.Path)
-		took := float64(duration.End(true).Total().Seconds())
-		m.Histogram.WithLabelValues(path, req.Method, strconv.Itoa(rw.code)).Observe(took)
+		// Only emit metrics on non-404 results, avoids metrics
+		// on arbitrary non existent paths.
+		if rw.code != http.StatusNotFound {
+			m.TSGauge.WithLabelValues(LabelEventSourceAPI, path, req.Method).Set(float64(duration.StartTime().Unix()))
+			took := float64(duration.End(true).Total().Seconds())
+			m.Histogram.WithLabelValues(path, req.Method, strconv.Itoa(rw.code)).Observe(took)
+		}
```

Comment:
@tommyp1ckles I might be changing my mind on this because a) the API is only accessible by a privileged client and b) there might be other http response codes that can be caused by arbitrary non existing paths, like 5xx for example. So the arbitrary non existent paths might not be a problem that we need to solve here.

Reply:
fixed